# TP 3 : Dynamic programmation applied to a knapsack problem

## Girard Antoine et Gonthier Priscilia


### Récupération des données

In [2]:
function readKnaptxtInstance(filename)
    price=[]
    weight=[]
    KnapCap=[]
    open(filename) do f
        for i in 1:3
            tok = split(readline(f))
            if(tok[1] == "ListPrices=")
                for i in 2:(length(tok)-1)
                    push!(price,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "ListWeights=")
                for i in 2:(length(tok)-1)
                    push!(weight,parse(Int64, tok[i]))
                end
            elseif(tok[1] == "Capacity=")
                push!(KnapCap, parse(Int64, tok[2]))
            else
                println("Unknown read :", tok)
            end 
        end
    end
    capacity=KnapCap[1]
    return price, weight, capacity
end

readKnaptxtInstance (generic function with 1 method)

### Fonction de résolution

In [3]:
function resolve(filename)
    price, weight, capacity = readKnaptxtInstance(filename)
    
    indices = sortperm(weight)
    
    sol = zeros(length(indices),capacity+1)
    
    # Calcul des éléments du tableau
    for i in 1:length(indices)
        for j in 1:capacity
            # Si i <= 1 : cas de la première ligne
            # Si on ne peut pas rentrer l'objet car le poids est supérieur à la capacité on laisse à 0
            # Sinon on met le coût de l'objet dans le tableau
            if i<=1 
                if j >= weight[indices[i]]
                    sol[i,j+1] = price[indices[i]]
                end
            # Si j-weight[indices[i]] < 0 : cas où le poids de l'objet i est supérieur à la capacité courante
            # On ne peut pas poser l'objet donc on prend le coût optimal précédent
            elseif j-weight[indices[i]] < 0
                sol[i,j+1] = sol[i-1,j+1]
            # Sinon : cas où la capacité courante est supérieure au poids de l'objet
            # Relation de récurrence
            # On prend la solution optimale entre :
            #    - prendre l'objet i : sol[i-1, j+1-weight[indices[i]]]+price[indices[i]]
            #    - ne pas prendre l'objet i : sol[i-1,j+1]
            else
                sol[i,j+1] = max(sol[i-1,j+1], sol[i-1, j+1-weight[indices[i]]]+price[indices[i]])
            end
        end  
    end
    
    BestProfit = sol[end,end]
    Bestsol = zeros(length(indices))
    j = capacity+1
    
    # Parcours inverse pour récupérer la solution.
    for i in length(indices):-1:1
        if i==1
            if sol[i,j] > 0
                Bestsol[indices[i]] = 1
            end
        elseif j-weight[indices[i]]>=0
            if sol[i-1,j]<sol[i,j]
                Bestsol[indices[i]] = 1
                j = j-weight[indices[i]]
            end       
        end
    end
    
    return BestProfit, Bestsol
end

resolve (generic function with 1 method)

### Affichage du résultat

In [4]:
BestProfit, Bestsol = resolve("instancesETU/KNAPnewformat/test.opb.txt")
println("\n******\n\nOptimal value = ", BestProfit, "\n\nOptimal x=", Bestsol)
#graphplot(trParentnodes, trChildnodes, names=trNamenodes, method=:tree)


******

Optimal value = 65.0

Optimal x=[0.0, 1.0, 0.0, 1.0]


#### Adéquation des résultats

On a effectué l'algorithme manuellement pour un sac à dos de capacité 10 et avec les objets : 

|       |       |       |       |       |
|:-:    |:-:    |:-:    |:-:    |:-:    |
| Poids | 3 | 4 | 5 | 7 |
| Prix | 12 |40 | 25 | 42 |  

et on obtient un profit maximal de 65 avec des objets de poids 4 et 5.  
C'est bien le résultat que l'on obtient avec notre algorithme de programmation dynamique.

## Test sur les instances du TP2

In [5]:
function test(name)
    fichier = "instancesETU/KNAPnewformat/" * name
    BestProfit, BestSol = resolve(fichier)
    println("Best Profit : ", BestProfit, "\nBest Solution : ", BestSol, "\n")
end

test (generic function with 1 method)

In [8]:
println("\n---------------------------------------")
println("50 données weakly_correlated")
println("profit attendu : 1396")
test("weakly_correlated/knapPI_2_50_1000_1_-1396.opb.txt")
println("\nprofit attendu : 13887")
test("weakly_correlated/knapPI_2_50_10000_1_-13887.opb.txt")

println("\n100 données weakly_correlated (l'execution du test était beaucoup trop longue pour le TP2)")
println("profit attendu : 343446")
test("weakly_correlated/knapPI_2_100_10000_5_-34346.opb.txt")

println("\n---------------------------------------")
println("\n50 données almost_strongly_correlated (l'execution du test était beaucoup trop longue pour le TP2)")
println("profit attendu : 2096")
test("almost_strongly_correlated/knapPI_5_50_1000_1_-2096.opb.txt")


println("\n---------------------------------------")
println("20 données circle")
println("profit attendu : 2291")
test("circle/knapPI_16_20_1000_1_-2291.opb.txt")

println("\n50 données circle")
println("profit attendu : 3408")
test("circle/knapPI_16_50_1000_1_-3408.opb.txt")


println("\n---------------------------------------")
println("50 données inverse_strongly_correlated")
println("profit attendu : 994")
test("inverse_strongly_correlated/knapPI_4_50_1000_1_-994.opb.txt")

println("\n100 données inverse_strongly_correlated")
println("profit attendu : 997")
test("inverse_strongly_correlated/knapPI_4_100_1000_1_-997.opb.txt")

println("\n200 données inverse_strongly_correlated (l'execution du test était beaucoup trop longue pour le TP2)")
println("profit attendu : 3152")
test("inverse_strongly_correlated/knapPI_4_200_1000_3_-3152.opb.txt")

println("\n---------------------------------------")
println("20 données multiple_strongly_correlated")
println("profit attendu : 1794")
test("multiple_strongly_correlated/knapPI_14_20_1000_1_-1794.opb.txt")

println("\n50 données multiple_strongly_correlated (l'execution du test était beaucoup trop longue pour le TP2)")
println("profit attendu : 3237")
test("multiple_strongly_correlated/knapPI_14_50_1000_5_-3237.opb.txt")

println("\n---------------------------------------")
println("20 données profit_ceiling")
println("profit attendu : 999")
test("profit_ceiling/knapPI_15_20_1000_1_-999.opb.txt")

println("\n50 données profit_ceiling (l'execution du test était beaucoup trop longue pour le TP2)")
println("profit attendu : 1047")
test("profit_ceiling/knapPI_15_50_1000_5_-1047.opb.txt")

println("\n---------------------------------------")
println("50 données similar_weights")
println("profit attendu : 995")
test("similar_weights/knapPI_9_50_1000_1_-995.opb.txt")

println("\n100 données similar_weights")
println("profit attendu : 995")
test("similar_weights/knapPI_9_100_1000_1_-995.opb.txt")

println("\n200 données similar_weights")
println("profit attendu : 995")
test("similar_weights/knapPI_9_200_1000_1_-995.opb.txt")


---------------------------------------
50 données weakly_correlated
profit attendu : 1396
Best Profit : 1396.0
Best Solution : [0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]


profit attendu : 13887
Best Profit : 13887.0
Best Solution : [0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]


100 données weakly_correlated (l'execution du test était beaucoup trop longue pour le TP2)
profit attendu : 343446
Best Profit : 45953.0
Best Solution : [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 

Best Profit : 5737.0
Best Solution : [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.

Pour la plupart des données testées, on obtient un profit maximal qui correspond à celui attendu  
Cependant nous avons un problème sur les données :   
weakly_correlated/knapPI_2_100_10000_5_-34346.opb.txt  
le profit trouvé est de 45953 au lieu de 34346  

multiple_strongly_correlated/knapPI_14_50_1000_5_-3237.opb.txt  
le profit trouvé est de 5737 au lieu de 3237  

profit_ceiling/knapPI_15_50_1000_5_-1047.opb.txt  
le profit trouvé est de 1062 au lieu de 1047  

 Nous avons donc créé une fonction pour déterminer si le profit trouvé est valide (c'est à dire correspond aux contraintes), si c'est le cas alors cela veut dire que la solution trouvée est correcte et que la solution attendue n'était pas la meilleure possible.

In [5]:
function verification(name)
    fichier = "instancesETU/KNAPnewformat/" * name
    price, weight, capacity = readKnaptxtInstance(fichier)
    BestProfit, BestSol = resolve(fichier)
    valide = true
    
    # Vérifier aue la capacité ne dépasse pas celle du sac à dos
    capacite_utilise = sum(BestSol .* weight)
    if capacite_utilise <= capacity
        println("La capacité utilisé est correcte : (capacité utilisée : ", capacite_utilise, ") <= (capacité totale : ", capacity,")")
    else
        println("La capacité utilisé est incorrecte : (capacité utilisée : ", capacite_utilise, ") > (capacité totale : ", capacity,")")
        valide = false
    end
    
    # Vérifier que le profit trouvé correspond au calculé
    profit = sum(price .* BestSol)
    if profit == BestProfit
        println("Le profit trouvé correspond au calculé : ", BestProfit)
    else
        println("Le profit trouvé (",BestProfit,") ne correspond pas au calculé (", profit, ")")
        valide = false
    end
    
    # Vérifier que l'on met bien chaque objet au maximum une fois dans le sac
    max_obj = maximum(BestSol)
    if max_obj > 1
        println("Un ou plusieurs objets sont mis plus d'une fois dans le sac (", max_obj, "fois)")
        valide = false
    else
        println("Chaque objet est mis au maximum une fois dans le sac")
    end
    
    println("Conclusion :La solution trouvée ", valide ? "est bien" : "n'est pas", " valide")
end

verification (generic function with 1 method)

In [6]:
println("Vérification pour le fichier weakly_correlated/knapPI_2_100_10000_5_-34346.opb.txt")
verification("weakly_correlated/knapPI_2_100_10000_5_-34346.opb.txt")

println("\nVérification pour le fichier multiple_strongly_correlated/knapPI_14_50_1000_5_-3237.opb.txt")
verification("multiple_strongly_correlated/knapPI_14_50_1000_5_-3237.opb.txt")

println("\nVérification pour le fichier profit_ceiling/knapPI_15_50_1000_5_-1047.opb.txt")
verification("profit_ceiling/knapPI_15_50_1000_5_-1047.opb.txt")

Vérification pour le fichier weakly_correlated/knapPI_2_100_10000_5_-34346.opb.txt
La capacité utilisé est correcte : (capacité utilisée : 24565.0) <= (capacité totale : 24600)
Le profit trouvé correspond au calculé : 45953.0
Chaque objet est mis au maximum une fois dans le sac
Conclusion :La solution trouvée est bien valide

Vérification pour le fichier multiple_strongly_correlated/knapPI_14_50_1000_5_-3237.opb.txt
La capacité utilisé est correcte : (capacité utilisée : 1037.0) <= (capacité totale : 1037)
Le profit trouvé correspond au calculé : 5737.0
Chaque objet est mis au maximum une fois dans le sac
Conclusion :La solution trouvée est bien valide

Vérification pour le fichier profit_ceiling/knapPI_15_50_1000_5_-1047.opb.txt
La capacité utilisé est correcte : (capacité utilisée : 1037.0) <= (capacité totale : 1037)
Le profit trouvé correspond au calculé : 1062.0
Chaque objet est mis au maximum une fois dans le sac
Conclusion :La solution trouvée est bien valide


Les solutions trouvés par l'algorithme de programmation dynamique sont donc bien toutes valides.

On obtient les mêmes valeurs de fonction-objectif pour le TP2. Cependant on remarque que l'execution du solver est plus rapide pour la programmation dynamique.